File: cdn-system/web/admin/src/services/guardPageService.ts
import type { GuardPageDefinition, GuardPageMap } from '@/types/guardPages'
import type { ErrorPageI18nSettings } from '@/types/errorPages'
import { DEFAULT_ERROR_PAGE_I18N } from '@/constants/errorPageLocales'
import guardDefaults from '@cdn-common/i18n/guard_default_strings.json'
import clickTemplate from '@cdn-common/i18n/guard_templates/click.html?raw'
import slideTemplate from '@cdn-common/i18n/guard_templates/slide.html?raw'
import captchaTemplate from '@cdn-common/i18n/guard_templates/captcha.html?raw'
import delayJumpTemplate from '@cdn-common/i18n/guard_templates/delay_jump.html?raw'
import rotateTemplate from '@cdn-common/i18n/guard_templates/rotate.html?raw'
import { extractTemplateKeys, normalizeLocaleTag } from '@/services/errorPageService'

const PLACEHOLDER_RE = /\{\{([a-zA-Z0-9_]+)\}\}/g

type GuardDefaultStrings = Record<string, Record<string, Record<string, string>>>

const DEFAULT_STRINGS = (guardDefaults as { strings: GuardDefaultStrings }).strings || {}

export const GUARD_PAGE_KEYS = ['click', 'slide', 'captcha', 'delay_jump', 'rotate'] as const

export type GuardPageKey = typeof GUARD_PAGE_KEYS[number]

export const ANTI_CC_TYPE_TO_PAGE_KEY: Record<string, GuardPageKey> = {
  slide: 'slide',
  slide_simple: 'slide',
  captcha: 'captcha',
  click: 'click',
  click_simple: 'click',
  '5s': 'delay_jump',
  rotate: 'rotate'
}

const DEFAULT_TEMPLATES: Record<GuardPageKey, string> = {
  click: clickTemplate,
  slide: slideTemplate,
  captcha: captchaTemplate,
  delay_jump: delayJumpTemplate,
  rotate: rotateTemplate
}

const mergeMissingStringValues = (
  target: Record<string, string>,
  fallback: Record<string, string> | undefined
): Record<string, string> => {
  const next = { ...target }
  if (!fallback) return next
  Object.entries(fallback).forEach(([key, value]) => {
    if (!String(next[key] || '').trim()) {
      next[key] = value
    }
  })
  return next
}

export const fillMissingGuardPageStrings = (
  pageKey: string,
  strings: Record<string, Record<string, string>> | undefined,
  langs: string[]
): Record<string, Record<string, string>> => {
  const next = { ...(strings || {}) }
  const pageDefaults = DEFAULT_STRINGS[pageKey] || {}
  langs.forEach(lang => {
    next[lang] = mergeMissingStringValues(next[lang] || {}, pageDefaults[lang])
  })
  Object.entries(pageDefaults).forEach(([lang, values]) => {
    next[lang] = mergeMissingStringValues(next[lang] || {}, values)
  })
  return next
}

export const resolveGuardPageKey = (antiCcType: string): GuardPageKey => {
  return ANTI_CC_TYPE_TO_PAGE_KEY[antiCcType] || 'click'
}

export const ensureGuardPageStructure = (
  pages: GuardPageMap | undefined,
  i18n: ErrorPageI18nSettings | undefined
): GuardPageMap => {
  const enabledLangs = Array.isArray(i18n?.enabled_langs) && i18n.enabled_langs.length
    ? [...new Set(i18n.enabled_langs.map(normalizeLocaleTag).filter(Boolean))]
    : [...DEFAULT_ERROR_PAGE_I18N.enabled_langs]

  const nextPages: GuardPageMap = {}
  GUARD_PAGE_KEYS.forEach(key => {
    const existing = pages?.[key]
    nextPages[key] = {
      template: existing?.template || DEFAULT_TEMPLATES[key] || '',
      strings: fillMissingGuardPageStrings(key, existing?.strings, enabledLangs)
    }
  })
  return nextPages
}

export const renderGuardPagePreview = (
  template: string,
  strings: Record<string, string>,
  lang = 'zh-CN'
): string => {
  if (!template) return ''
  const withStrings = template.replace(PLACEHOLDER_RE, (_, key) => {
    if (key === 'html_lang') {
      return lang
    }
    return strings?.[key] ?? `{{${key}}}`
  })
  return withStrings
}

export const resolveGuardPreviewStrings = (
  def: GuardPageDefinition,
  lang: string,
  defaultLang: string
): Record<string, string> => {
  const candidates = [lang, lang.split('-')[0], defaultLang, 'zh-CN', 'en']
  for (const candidate of candidates) {
    const strings = candidate ? def.strings?.[candidate] : undefined
    if (strings) {
      return strings
    }
  }
  const first = Object.values(def.strings || {})[0] as Record<string, string> | undefined
  return first || {}
}

export const extractGuardTemplateKeys = (template: string): string[] => extractTemplateKeys(template)

export const migrateLegacyAntiCCPageCustom = (
  guardPages: GuardPageMap,
  antiCcType: string,
  legacyCustom: string
): GuardPageMap => {
  const custom = String(legacyCustom || '').trim()
  if (!custom) {
    return guardPages
  }
  const pageKey = resolveGuardPageKey(antiCcType)
  const page = guardPages[pageKey]
  if (page?.template && page.template !== DEFAULT_TEMPLATES[pageKey]) {
    return guardPages
  }
  return {
    ...guardPages,
    [pageKey]: {
      ...(page || { template: DEFAULT_TEMPLATES[pageKey], strings: {} }),
      template: custom
    }
  }
}

File: cdn-system/api/services/guard_page_service.go
package services

import (
	"encoding/json"
	"fmt"
	"strings"

	"cdn-api/models"
	"cdn-common/i18n"
)

var guardPageKeys = i18n.GuardPageKeys()

var antiCCTypeToPageKey = map[string]string{
	"slide":        "slide",
	"slide_simple": "slide",
	"captcha":      "captcha",
	"click":        "click",
	"click_simple": "click",
	"5s":           "delay_jump",
	"rotate":       "rotate",
}

var guardPageTemplateFiles = map[string]string{
	"click":      "click.html",
	"slide":      "slide.html",
	"captcha":    "captcha.html",
	"delay_jump": "delay_jump.html",
	"rotate":     "rotate.html",
}

func GuardPageKeyForAntiCCType(antiCCType string) string {
	key := strings.TrimSpace(antiCCType)
	if mapped, ok := antiCCTypeToPageKey[key]; ok {
		return mapped
	}
	return "click"
}

func DefaultGuardPageDefinitions() map[string]models.GuardPageDefinition {
	defaultStrings := i18n.GuardDefaultStrings()
	defaultTemplates := i18n.GuardDefaultTemplates()
	enabledLangs := DefaultErrorPageI18nSettings().EnabledLangs
	pages := make(map[string]models.GuardPageDefinition, len(guardPageKeys))
	for _, key := range guardPageKeys {
		pages[key] = models.GuardPageDefinition{
			Template: defaultTemplates[key],
			Strings:  fillMissingGuardPageStrings(key, nil, enabledLangs, defaultStrings),
		}
	}
	return pages
}

func fillMissingGuardPageStrings(
	pageKey string,
	strings map[string]map[string]string,
	langs []string,
	defaults map[string]map[string]map[string]string,
) map[string]map[string]string {
	next := map[string]map[string]string{}
	for lang, values := range strings {
		next[lang] = copyStringMap(values)
	}
	pageDefaults := defaults[pageKey]
	for _, lang := range langs {
		next[lang] = mergeMissingStringValues(next[lang], pageDefaults[lang])
	}
	for lang, values := range pageDefaults {
		next[lang] = mergeMissingStringValues(next[lang], values)
	}
	return next
}

func mergeMissingStringValues(target map[string]string, fallback map[string]string) map[string]string {
	out := copyStringMap(target)
	for key, value := range fallback {
		if strings.TrimSpace(out[key]) == "" {
			out[key] = value
		}
	}
	return out
}

func copyStringMap(input map[string]string) map[string]string {
	if len(input) == 0 {
		return map[string]string{}
	}
	out := make(map[string]string, len(input))
	for key, value := range input {
		out[key] = value
	}
	return out
}

func NormalizeGlobalConfigGuardPages(cfg *models.GlobalConfig) {
	if cfg == nil {
		return
	}
	NormalizeGlobalConfigErrorPages(cfg)
	enabledLangs := cfg.ErrorPageI18n.EnabledLangs
	if len(enabledLangs) == 0 {
		enabledLangs = DefaultErrorPageI18nSettings().EnabledLangs
	}
	defaults := DefaultGuardPageDefinitions()
	if cfg.GuardPages == nil {
		cfg.GuardPages = map[string]models.GuardPageDefinition{}
	}
	for _, key := range guardPageKeys {
		existing := cfg.GuardPages[key]
		def := defaults[key]
		if strings.TrimSpace(existing.Template) == "" {
			existing.Template = def.Template
		}
		existing.Strings = fillMissingGuardPageStrings(key, existing.Strings, enabledLangs, i18n.GuardDefaultStrings())
		cfg.GuardPages[key] = existing
	}
	migrateLegacyAntiCCPageCustom(cfg)
}

func migrateLegacyAntiCCPageCustom(cfg *models.GlobalConfig) {
	custom := strings.TrimSpace(cfg.WAF.AntiCCPageCustom)
	if custom == "" {
		return
	}
	pageKey := GuardPageKeyForAntiCCType(cfg.WAF.AntiCCType)
	page := cfg.GuardPages[pageKey]
	if strings.TrimSpace(page.Template) != "" && page.Template != i18n.GuardDefaultTemplates()[pageKey] {
		return
	}
	page.Template = custom
	cfg.GuardPages[pageKey] = page
	cfg.WAF.AntiCCPageCustom = ""
}

func ValidateGuardPageConfig(cfg *models.GlobalConfig) error {
	if cfg == nil {
		return nil
	}
	NormalizeGlobalConfigGuardPages(cfg)
	for _, key := range guardPageKeys {
		page := cfg.GuardPages[key]
		if strings.TrimSpace(page.Template) == "" {
			return fmt.Errorf("guard page %s template is required", key)
		}
		keys := extractTemplateKeys(page.Template)
		if len(keys) == 0 {
			continue
		}
		defaultLang := cfg.ErrorPageI18n.DefaultLang
		if defaultLang == "" {
			defaultLang = "zh-CN"
		}
		stringsForLang := page.Strings[defaultLang]
		if stringsForLang == nil {
			stringsForLang = map[string]string{}
		}
		for _, placeholder := range keys {
			if placeholder == "html_lang" {
				continue
			}
			if strings.TrimSpace(stringsForLang[placeholder]) == "" {
				return fmt.Errorf("guard page %s missing default language string for {{%s}}", key, placeholder)
			}
		}
	}
	return nil
}

func GuardPageTemplateFileName(pageKey string) string {
	if file, ok := guardPageTemplateFiles[pageKey]; ok {
		return file
	}
	return pageKey + ".html"
}

func BuildGuardI18nPayload(pages map[string]models.GuardPageDefinition) map[string]interface{} {
	strings := map[string]map[string]map[string]string{}
	for _, key := range guardPageKeys {
		page := pages[key]
		if len(page.Strings) == 0 {
			continue
		}
		strings[key] = page.Strings
	}
	return map[string]interface{}{
		"strings": strings,
	}
}

func parseGuardPagesFromRaw(raw json.RawMessage) map[string]models.GuardPageDefinition {
	if len(raw) == 0 {
		return nil
	}
	var pages map[string]models.GuardPageDefinition
	if err := json.Unmarshal(raw, &pages); err != nil {
		return nil
	}
	return pages
}
